### PR TITLE
feat: WASM extras

### PR DESCRIPTION
**Pull request by danwilliams**
- Added `storage_remove()`, which was missing.

  - Added a `generate_uuid()` host function. This is necessary because the guest does not have randomness available.

  - Added a `time_now()` host function. This is necessary because the guest does not have randomness available.

### CODE REVIEW

File: crates/runtime/Cargo.toml
Comment (miraclx) [inline crates/runtime/Cargo.toml]: these directives are no longer necessary

File: crates/runtime/src/logic.rs
Comment (miraclx) [inline crates/runtime/src/logic.rs]: there's no fundamental reason to have a non-mut logic reference in this case, can be simplified to

```suggestion
        self.with_logic_mut(|logic| {
            let uuid = logic.generate_uuid();
```

File: crates/runtime/src/logic.rs
Comment (miraclx) [inline crates/runtime/src/logic.rs]: although, there's also no reason to have the `VMLogic::generate_uuid` convenience method since this specialized functionality will only be used by this specialized function

```suggestion
        let uuid = Uuid::new_v4();
        self.with_logic_mut(|logic| {
```

File: crates/runtime/src/logic.rs
Comment (miraclx) [inline crates/runtime/src/logic.rs]: same here

```suggestion
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Time went backwards to before the Unix epoch!")
            .as_nanos() as u64
```

File: crates/runtime/src/logic/imports.rs
Comment (miraclx) [inline crates/runtime/src/logic/imports.rs:64]: `generate_uuid` can be further generalized to `random_bytes::<N>` that returns random bytes of arbitrary length

Knowledge of the fact that it's a `Uuid` stays with storage, neither the SDK nor runtime need that information, and can be constructed by storage from a `[u8; 16]` via [`uuid::Builder::from_random_bytes`](https://docs.rs/uuid/latest/uuid/struct.Builder.html#method.from_random_bytes)

File: crates/sdk/Cargo.toml
Comment (miraclx) [inline crates/sdk/Cargo.toml]: ^

File: crates/runtime/src/logic.rs
Comment (miraclx) [inline crates/runtime/src/logic.rs]:
```suggestion
                .set(logic.limits, register_id, uuid.into_bytes())
```

File: crates/sdk/src/env.rs
Comment (miraclx) [inline crates/sdk/src/env.rs:201]: these are not necessary, anything that would use `generate_uuid` would at least have already called `state_read -> storage_read`, which would according to https://github.com/calimero-network/core/blob/54336ece92deffa11447c54e1f420fb959f7d727/crates/sdk/src/sys.rs#L48-L50 simply say "not supported"

the same would be maintained of `sys::generate_uuid`, so the mocks are not needed

File: crates/sdk/src/env.rs
Comment (miraclx) [inline crates/sdk/src/env.rs:229]: same here

File: crates/storage/Cargo.toml
Comment (miraclx) [inline crates/storage/Cargo.toml]: ^

**Review by miraclx — CHANGES_REQUESTED** [review]

File: crates/runtime/Cargo.toml
Comment (danwilliams) [reply crates/runtime/Cargo.toml]: Could you explain? 🙂 At present these are here to ensure that we can achieve the same functionality both through WASM and not. If these were removed, what would we do instead?

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: That's a fair observation 👍 Done! 🙂

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: Ahhhh - no; this is actually used a little more widely. I think it's best we keep this calling `generate_uuid()` as that is then consistent with the rest of the usage.

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: Agreed - calling `logic.time_now()` here is better. Additionally, in this location we can indeed subsume the `logic.time_now()` call into this location, as it's the only time we use it. So although it was added to try to preserve consistency with the other, similar implementations, there is no special need for a separate function here. So this has been removed 🙂 👍

File: crates/runtime/src/logic/imports.rs
Comment (danwilliams) [reply crates/runtime/src/logic/imports.rs:64]: That's an interesting thought. Yes, that could potentially be done, and I can see your reasoning for that. However, in this situation we don't have a need for arbitrary length, and also this would not fit neatly with the UUID functionality. It's definitely a potential for later, but I've left this as-is for now 🙂

File: crates/sdk/Cargo.toml
Comment (danwilliams) [reply crates/sdk/Cargo.toml]: I will need to know a little more about your intentions here 🙂

File: crates/sdk/src/env.rs
Comment (danwilliams) [reply crates/sdk/src/env.rs:229]: So I think the purpose here has been misunderstood... and the premise seems incorrect. If I remove this, then the build script no longer works:

```
error[E0432]: unresolved import `calimero_sdk::env::time_now`
   --> crates/storage/src/entities.rs:219:5
    |
219 | use calimero_sdk::env::time_now;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `time_now` in `env`
    |
note: found an item that was configured out
   --> /Users/dan/Code/calimero/crates/sdk/src/env.rs:233:8
    |
233 | pub fn time_now() -> u64 {
    |        ^^^^^^^^
note: the item is gated behind the `wasm32` feature
   --> /Users/dan/Code/calimero/crates/sdk/src/env.rs:230:7
    |
230 | #[cfg(target_arch = "wasm32")]
    |       ^^^^^^^^^^^^^^^^^^^^^^
```

It may be possible to do what you're suggesting in a different way, but I'm not sure how, and what's in place currently works.

File: crates/sdk/src/env.rs
Comment (danwilliams) [reply crates/sdk/src/env.rs:201]: See below example with `time_now()`, as that's unaffected by the wider question around UUID type. For this specific context, the same applies - i.e. cannot remove this.

File: crates/runtime/Cargo.toml
Comment (miraclx) [reply crates/runtime/Cargo.toml]: they're the same thing, just merge it into `[dependencies]`

File: crates/runtime/src/logic.rs
Comment (miraclx) [reply crates/runtime/src/logic.rs]: > this is actually used a little more widely

by?

File: crates/runtime/src/logic.rs
Comment (miraclx) [reply crates/runtime/src/logic.rs]: > Additionally, in this location we can indeed subsume the logic.time_now() call into this location

not necessarily, I like to think of these like `unsafe {}` (it technically is behind the scenes too), limit the scope of the closure to the use of `logic` and time doesn't use `logic`, so it can be outside

the same way unsafe things should be outside unsafe blocks even though unsafe things use their outputs

File: crates/runtime/src/logic/imports.rs
Comment (miraclx) [reply crates/runtime/src/logic/imports.rs:64]: The concern is, by the time this ships, we'll have a contract to uphold between built developer apps and the interfaces the runtime supports, where there's no reason to.

The ideal approach to designing a bridge is to be liberal within boundaries defined with good reason, and in this case, the generic approach doesn't incur any extra cost.

Randomness has always been a feature we wanted to introduce, it just had to be done safely which is only possible by optimistic execution. As this is the effort that introduces optimistic execution, and a component of its - `uuid` needs randomness, then we have to provide the fundamental facility for randomness.

Otherwise, what's likely to happen after this merges as is, is developers need random numbers and abuse `generate_uuid` to get that. Calling `generate_uuid` N times to get `16N` random bytes and not using it at all for uuids.

File: crates/sdk/Cargo.toml
Comment (miraclx) [reply crates/sdk/Cargo.toml]: same as the snippet in `runtime`, this directive can be merged into `[dependencies]`, although following https://github.com/calimero-network/core/pull/799#discussion_r1800168580, it won't be necessary to have the dependency in the SDK at all

File: crates/sdk/src/env.rs
Comment (miraclx) [reply crates/sdk/src/env.rs:229]: I'm confused, the point of this comment is to highlight that feature gating for wasm32 is unnecessary.

File: crates/sdk/src/env.rs
Comment (miraclx) [reply crates/sdk/src/env.rs:201]: Replied below

File: crates/storage/Cargo.toml
Comment (miraclx) [reply crates/storage/Cargo.toml]: move to `[dependencies]`, unless you actually need that `v5` feature activated, in which case, I can't find any examples of it's use

File: crates/runtime/Cargo.toml
Comment (danwilliams) [reply crates/runtime/Cargo.toml]: Ahhh, I see. Yes - I had overlooked that in this location they ended up being the same. Well-spotted!

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: If you search the codebase for the function, you'll see it's used in a fair few places 🙂

File: crates/sdk/Cargo.toml
Comment (danwilliams) [reply crates/sdk/Cargo.toml]: Ahhh, I see. No - they are not the same thing. The WASM code cannot have the v4. In the other snippet, they were indeed the same, and I had overlooked that they had become so. Which means, that was a valid observation. But here, they are not the same.

File: crates/sdk/src/env.rs
Comment (danwilliams) [reply crates/sdk/src/env.rs:229]: Yes, I understand the point of the comment, but as illustrated above, that's incorrect. If you attempt what you have described, then the build fails. Therefore this is necessary.

File: crates/storage/Cargo.toml
Comment (danwilliams) [reply crates/storage/Cargo.toml]: As per above, they are not the same - the WASM cannot have the v4.

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: > > Additionally, in this location we can indeed subsume the logic.time_now() call into this location
> 
> not necessarily

I am confused - this was your suggestion? 😄 I was agreeing with you that this could indeed happen here, and I've made that change.

If the change is incorrect, can you explain what you intend more clearly?

If the focus is about the logic, then the move inside agrees with your comments.

If the focus is about what's unsafe or not, then that's the way I had previously approached it, in both locations.

File: crates/runtime/src/logic/imports.rs
Comment (danwilliams) [reply crates/runtime/src/logic/imports.rs:64]: Hmmmm, I am not sure that there is actually any effect from that, i.e. I do not think the outcome varies.

Let's consider: In a situation where you ask for random bytes, returning `[u8; 16]`, then you get random bytes and can do that as many times as you like. Ergo, the situation you describe as being undesirable. It's the same outcome.

However, in this case we are using semantic identification to recognise that this is a UUID. So it's more constrained than a random number generator of arbitrary length, but also is specific to a purpose of providing UUIDs.

Yes they could call it umpteen times - but they could do this in both cases. What's the difference?

Therefore, on balance, it seems better to retain this as a UUID type, because:

  - There is no difference in randomness or ability to call it multiple times
  - There is semantic identification of the type in use
  - The approach taken to generate a UUID is a specific algorithm and not directly equivalent to an arbitrary randomness generator
  - The exact approach we take to generate UUIDs is abstracted away, and we are free to change it in future without it being tied to arbitrary randomness

Hence it would appear that using a UUID here, i.e. the `Uuid` type, is the correct choice - and even more importantly, that using the UUID functionality for generation and not an arbitrary randomness generator is also correct.

From a contract perspective, the only contract we are formally providing is that an `Id` performs in a particular way. We are free to change `Id` internals later without breaking that contract.

File: crates/runtime/src/logic.rs
Comment (miraclx) [reply crates/runtime/src/logic.rs]: I did, and I couldn't find any references, which is why I asked, `VMLogic` isn't to be used by anything but `calimero_runtime::run` anyway, so it can't be used by anything else

File: crates/runtime/src/logic.rs
Comment (miraclx) [reply crates/runtime/src/logic.rs]: > If the change is incorrect
> If the focus is about the logic, then the move inside agrees with your comments.

so `logic` is just a variable, `with_logic_mut` is a safe way of fetching it because it's self-referencing and that requires `unsafe`

so it's not incorrect, it just makes sense to limit the scope of the closure to things that _actually need to use the variable_

File: crates/runtime/src/logic/imports.rs
Comment (miraclx) [reply crates/runtime/src/logic/imports.rs:64]: > In a situation where you ask for random bytes, returning [u8; 16], then you get random bytes and can do that as many times as you like. Ergo, the situation you describe as being undesirable. It's the same outcome.

as mentioned earlier:

> `generate_uuid` can be further generalized to `random_bytes::<N>` that returns random bytes of arbitrary length

so it's not the same

> The approach taken to generate a UUID is a specific algorithm and not directly equivalent to an arbitrary randomness generator

there is no "specific algorithm" for UUID v4, the only requirement is you have sufficiently random bytes, and appropriately set the version and variant when making those random bytes into a UUID

> The exact approach we take to generate UUIDs is abstracted away, and we are free to change it in future without it being tied to arbitrary randomness

ideally, if you don't trust a randomness source to be cryptographically secure, you'd pass that as a seed to a CSPRNG like ChaCha. That's what [`StdRng`](https://docs.rs/rand/latest/rand/rngs/struct.StdRng.html) does for example.

> From a contract perspective, the only contract we are formally providing is that an Id performs in a particular way. We are free to change Id internals later without breaking that contract.

that's maintained

File: crates/sdk/Cargo.toml
Comment (miraclx) [reply crates/sdk/Cargo.toml]: it doesn't need the `v5` either, it just needs to call `generate_uuid` which returns a UUID, so you don't need this directive

just make it

```toml
[dependencies]
uuid.workspace = true
```

but IMO (https://github.com/calimero-network/core/pull/799#discussion_r1800168580), this dependency doesn't even need to exist in the SDK at all

File: crates/sdk/src/env.rs
Comment (miraclx) [reply crates/sdk/src/env.rs:229]: The build fails because you still have a feature flag 🤷🏽‍♂️

File: crates/storage/Cargo.toml
Comment (miraclx) [reply crates/storage/Cargo.toml]: Replied above, it doesn't need to, yet this still stands

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: Okay, so does that mean that this is resolved...? I am not sure 😅

File: crates/runtime/src/logic/imports.rs
Comment (danwilliams) [reply crates/runtime/src/logic/imports.rs:64]: > there is no "specific algorithm" for UUID v4, the only requirement is you have sufficiently random bytes, and appropriately set the version and variant when making those random bytes into a UUID

That's self-contradicting 🙂 I'm glad you agree with me that v4 UUIDs are not the same as a set of random bytes. As you've described, they have a version and variant built in, which makes them functionally different.

I believe there's no change to make here.

File: crates/runtime/src/logic.rs
Comment (miraclx) [reply crates/runtime/src/logic.rs]: Localizing the call to this location without involving `VMLogic`, yes.

However, the `let now = _` statement should be outside the closure.

File: crates/runtime/src/logic/imports.rs
Comment (miraclx) [reply crates/runtime/src/logic/imports.rs:64]: > That's self-contradicting 🙂 I'm glad you agree with me that v4 UUIDs are not the same as a set of random bytes. As you've described, they have a version and variant built in, which makes them functionally different.

it's not, as mentioned earlier, to construct a v4 uuid with self-provided random bytes, you'd use [`uuid::Builder::from_random_bytes`](https://docs.rs/uuid/latest/uuid/struct.Builder.html#method.from_random_bytes) which proceeds to set the version and variant.

Everything else is random. That's what matters.

File: crates/sdk/Cargo.toml
Comment (danwilliams) [reply crates/sdk/Cargo.toml]: It's not possible to do `uuid.workspace = true` here, as it doesn't exist in the workspace dependencies - it can't, due to this exact issue of arch targeting.

No problem removing the v5, although that was consciously there for use in tests, but that's not strictly necessary and not currently used.

File: crates/sdk/src/env.rs
Comment (danwilliams) [reply crates/sdk/src/env.rs:229]: That's the point, isn't it? In order to build, this needs to be available 🤔

File: crates/runtime/src/logic.rs
Comment (danwilliams) [reply crates/runtime/src/logic.rs]: Thank you for confirming 🙂

File: crates/runtime/src/logic/imports.rs
Comment (danwilliams) [reply crates/runtime/src/logic/imports.rs:64]: In order to do that, you still need to expose a source of random bytes, though, even if you are then feeding them into UUID.

Initially, I thought you were talking about returning a `[u8; 16]` of random bytes. I objected to that for obvious reasons.

However, if you are saying you are constructing an actual UUID type then you still need that type available, and you also still need a source of randomness.

I don't see any change to make here - what is already present works, and does what it should, and I'm not sure I even understand fully what you are intending to do, as your objectives are not totally clear to me.

File: crates/sdk/Cargo.toml
Comment (miraclx) [reply crates/sdk/Cargo.toml]: > It's not possible to do uuid.workspace = true here, as it doesn't exist in the workspace dependencies

it can be introduced

> it can't, due to this exact issue of arch targeting.

not true, with the version specified in the root manifest, you could do this in the local crate's

```toml
[target.'cfg(target_arch = "wasm32")'.dependencies]
uuid = { workspace = true, features = ["serde", "v5"] }

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
uuid = { workspace = true, features = ["serde", "v4"] }
```

File: crates/sdk/src/env.rs
Comment (miraclx) [reply crates/sdk/src/env.rs:229]: and that's the point, it doesn't need it, you don't need to accommodate for anything but wasm

File: crates/runtime/src/logic/imports.rs
Comment (miraclx) [reply crates/runtime/src/logic/imports.rs:64]: > In order to do that, you still need to expose a source of random bytes, though, even if you are then feeding them into UUID.

that's the idea, as mentioned in the original comment

> However, if you are saying you are constructing an actual UUID type then you still need that type available, and you also still need a source of randomness.

as mentioned, not only are we NOT constructing a UUID type at this level, we do not need to even have that dependency in the runtime nor SDK

> I don't see any change to make here

`random_bytes::<N>` for runtime & SDK, `uuid::Builder::from_random_bytes` for storage

> and I'm not sure I even understand fully what you are intending to do, as your objectives are not totally clear to me.

https://github.com/calimero-network/core/pull/799#discussion_r1800312350, not sure how else to explain

Comment (danwilliams) [reply crates/runtime/src/logic/imports.rs:64]: Okay, so I think I should recap my previous points:

> Therefore, on balance, it seems better to retain this as a UUID type, because:
> 
>   - There is no difference in randomness or ability to call it multiple times
>   - There is semantic identification of the type in use
>   - The approach taken to generate a UUID is a specific algorithm and not directly equivalent to an arbitrary randomness generator
>   - The exact approach we take to generate UUIDs is abstracted away, and we are free to change it in future without it being tied to arbitrary randomness
> 
> Hence it would appear that using a UUID here, i.e. the Uuid type, is the correct choice - and even more importantly, that using the UUID functionality for generation and not an arbitrary randomness generator is also correct.

I think the best path here is to merge this and then any subsequent changes can be carried out as you see fit - I do not see any harm in UUID being available; the user is able to do the same thing in both cases, and having the actual type present is important and beneficial.

File: crates/sdk/Cargo.toml
Comment (danwilliams) [reply crates/sdk/Cargo.toml]: Okay, so that's fair about introducing it into the main workspace - my preference was not to do that, as I like it being obvious to have the arch matches together, but it's also fine having it in the main list and then overridden 👍

File: crates/sdk/src/env.rs
Comment (danwilliams) [reply crates/sdk/src/env.rs:229]: For whatever reason, if that is removed, the build script fails.

Comment (miraclx) [reply crates/sdk/src/env.rs:229]: what does it look like when it's removed?

File: Cargo.toml
Comment (miraclx) [inline Cargo.toml:120]: the root manifest should just define versions which can be inherited with no default feature activations unless with good reason (security or other similar constraint)

File: Cargo.toml
Comment (miraclx) [inline Cargo.toml]: we don't need this in the root

File: crates/sdk/src/env.rs
Comment (danwilliams) [reply crates/sdk/src/env.rs:229]: You know more about this side than I do - I've only recently started fiddling with it, in order to try to test this stuff. I wrote what I did in order for it to fully work. Removing what you've suggested breaks it in the way described. I'd say the best path forward at this stage is for you to make the changes that you are thinking of, as it may be that I've misunderstood you.

File: Cargo.toml
Comment (danwilliams) [reply Cargo.toml:120]: The root manifest here defines the features that all usages of this crate need, i.e. serde. That is used everywhere that the crate is used, and is in place as a default.

Comment (fbozic) [reply Cargo.toml:120]: I agree with mira on this one, root manifest should just define versions

Comment (fbozic) [reply Cargo.toml:120]: Could you also clean up `fixedstr`, `uuid` and `axum-server` deps? I just saw they are not following the pattern. Thanks

**Review by miraclx — APPROVED** [review]
approving to unblock, pending comments will be resolved in coming patches

**Pull request merged**
